### PR TITLE
Added code

### PR DESCRIPTION
**Pull request by cyrusHwang**

### CODE REVIEW

File: progression-activity-analysis.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+for i in d.items():
+    total_incidents += i[1][1]
+    total_ffa += i[1][0]
+    print("User {}: For sure cheated: {} Failed first attempts: {}".format(i[0], i[1][1], i[1][0]))
+    
+    
+print('\nTotal incidents = {}'.format(total_incidents))
+print('Total failed first attempts = {}'.format(total_ffa))
+
+
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: Bunch of blank lines. Please remove

File: progression-activity-analysis.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                    if (typeTupleList(rec[3]) != answerFormat or isRepeat(rec[3])):
+                        d[rec[1]][1] += 1
+
+total_incidents = 0
+total_ffa = 0
+for i in d.items():
+    total_incidents += i[1][1]
+    total_ffa += i[1][0]
+    print("User {}: For sure cheated: {} Failed first attempts: {}".format(i[0], i[1][1], i[1][0]))
+
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: Extra line, can remove

File: progression-activity-analysis.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+def typeTupleList(string):
+    return [(c.isnumeric(), c.isalpha(), not c.isalnum()) for c in string.replace('-', '').replace('+', '').split(',')]
+
+def isRepeat(string):
+    l = string.split(',')
+    return len(l) > 1 and all(l[i] == l[0] for i in range(1, len(l)))
+
+myPath = "C:/Users/hwang/Programming/Work/folder1"
+f = open('output1.txt', 'a')
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: Please use more descriptive variable name

File: progression-activity-analysis.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import os
+
+def typeTupleList(string):
+    return [(c.isnumeric(), c.isalpha(), not c.isalnum()) for c in string.replace('-', '').replace('+', '').split(',')]
+
+def isRepeat(string):
+    l = string.split(',')
+    return len(l) > 1 and all(l[i] == l[0] for i in range(1, len(l)))
+
+myPath = "C:/Users/hwang/Programming/Work/folder1"
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: Anyway to make call to get current path? Maybe in os library?

File: progression-activity-analysis.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+import csv
+from collections import defaultdict
+import os
+
+def typeTupleList(string):
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: Please comment what this function does

Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: Ex: 
```
'''
What this function does....
'''
```

File: progression-activity-analysis.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    return [(c.isnumeric(), c.isalpha(), not c.isalnum()) for c in string.replace('-', '').replace('+', '').split(',')]
+
+def isRepeat(string):
+    l = string.split(',')
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: Please use more descriptive variable name. Ex: `csv_list`

File: progression-activity-analysis.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    return len(l) > 1 and all(l[i] == l[0] for i in range(1, len(l)))
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: Probably good to pull out `l[i] == l[0] for i in range(1, len(l))` and assign to meaningful variable name

File: progression-activity-analysis.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
+import csv
+from collections import defaultdict
+import os
+
+def typeTupleList(string):
+    return [(c.isnumeric(), c.isalpha(), not c.isalnum()) for c in string.replace('-', '').replace('+', '').split(',')]
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: Probably possible to make more clear what this is doing and why. Ex: Using a few more lines of code, or showing an explicit example

File: progression-activity-analysis.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+def isRepeat(string):
+    l = string.split(',')
+    return len(l) > 1 and all(l[i] == l[0] for i in range(1, len(l)))
+
+myPath = "C:/Users/hwang/Programming/Work/folder1"
+f = open('output1.txt', 'a')
+
+d = defaultdict(lambda : [0,0])
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: Please use more descriptive variable name than `d`

File: progression-activity-analysis.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        reader = csv.reader(csvfile)
+        next(reader)
+        for rec in reader:
+            if rec[2] == "1":
+                answerFormat = typeTupleList(rec[3])
+                break
+    with open(myPath + '/' + filename, 'r') as csvfile:
+        reader = csv.reader(csvfile)
+        next(reader)
+        for rec in reader:
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: This could use an explanation comment behind the rationale

**Comment by AlexEdgcomb** [comment]
@cyrusHwang , I've made first pass. Please see comments.

File: progression-activity-analysis.py (hunk 2 above)
Comment (cyrusHwang) [inline progression-activity-analysis.py]: Removed other line.

File: progression-activity-analysis.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+'''
+dictionary whose keys are userIDs and values are lists whose 1st
+index is the total number of failed first attempts and 2nd index
+is the total number of activities user has done
+'''
+userFirstTimeCorrectness = defaultdict(lambda : [0,0])
+
+for filename in os.listdir(folderPath):
+    s = set()
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: `s` is a set of users? Please use a more meaningful variable name

File: progression-activity-analysis.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    with open(folderPath + '/' + filename, 'r') as csvfile:
+        reader = csv.reader(csvfile)
+        next(reader)
+        for rec in reader:
+            if (rec[1] not in s):
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: What is rec[1]? Perhaps assign rec[1] to a variable to make clear what rec[1] is.

File: progression-activity-analysis.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+                userFirstTimeCorrectness[rec[1]][1] += 1
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: What is `rec[1][1]`? Perhaps assign to var to make more clear

Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: Oops. I see this isn't `rec[1][1]`. What's `userFirstTimeCorrectness[...][1]` though ?

File: progression-activity-analysis.py [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+                s.add(rec[1])
+                if rec[2] == "0":
+                    userFirstTimeCorrectness[rec[1]][0] += 1
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: What's `userFirstTimeCorrectness[...][0]` ?

File: progression-activity-analysis.py [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+outputFile.write("<User ID>: <Percentage Incorrect First Attempts>, <Total Activities>\n")
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: Please output as a CSV (comma separated value). Ex: `User ID,Percentage incorrect first time,Total activities\n`

File: progression-activity-analysis.py [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+        for rec in reader:
+            if (rec[1] not in s):
+                userFirstTimeCorrectness[rec[1]][1] += 1
+                s.add(rec[1])
+                if rec[2] == "0":
+                    userFirstTimeCorrectness[rec[1]][0] += 1
+
+outputFile.write("<User ID>: <Percentage Incorrect First Attempts>, <Total Activities>\n") 
+for k, v in sorted(userFirstTimeCorrectness.items(), key=lambda x : (x[1][0], x[1][1]), reverse=True):
+    outputFile.write("{}: {:.2f}%, {}\n".format(k, 100 * v[0]/v[1], v[1]))
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: Similarly, please output as CSV

**Comment by cyrusHwang** [comment]
@AlexEdgcomb

File: progression-activity-analysis.py [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+            if (userID not in usersInActivity):
+                userFirstTimeCorrectness[userID][1] += 1
+                usersInActivity.add(userID)
+                if answerIsCorrect == "0":
+                    userFirstTimeCorrectness[userID][0] += 1
+
+with open("data.csv", "w", newline='') as outputCsvfile:
+    writer = csv.writer(outputCsvfile) 
+    writer.writerow(["User ID", "Percentage Incorrect First Attempts", "Total Activities Participated"]) 
+    for userID, [numIncorrectFirstAttempts, participated] in sorted(userFirstTimeCorrectness.items(), key=lambda x : (x[1][0], x[1][1]), reverse=True):
```
Comment (AlexEdgcomb) [inline progression-activity-analysis.py]: Please make more clear what `sorted(userFirstTimeCorrectness.items(), key=lambda x : (x[1][0], x[1][1]), reverse=True)` is doing. Perhaps split the code over 2-3 lines, using variable names to document meaning.

Also, please give `x` a meaningful name. Thanks!

**Comment by AlexEdgcomb** [comment]
@cyrusHwang , just 1 more comment.

**Comment by cyrusHwang** [comment]
@AlexEdgcomb

**Comment by AlexEdgcomb** [comment]
@cyrusHwang , nice work. I've merged this branch into master.

You'll need to update your local master branch:
`git pull origin master`

Then, you can delete this PR's branch (after checking out master):
```
git checkout master
git branch -d firstVersion
```

After the above, when you run `git branch` there should only be `master`.

Slack me for help if needed.

**Pull request merged**